src/cargo/core/compiler/custom_build.rs

```diff
@@ -679,47 +683,120 @@ impl BuildOutput {
+
+        fn parse_directive<'a>(
+            whence: &str,
+            line: &str,
+            data: &'a str,
+            new_syntax: bool,
+            has_metadata_prefix: bool,
+        ) -> CargoResult<(&'a str, &'a str)> {
+            let mut iter = data.splitn(2, "=");
+            let key = iter.next();
+            let value = iter.next();
+            match (key, value) {
+                (Some(a), Some(b)) => Ok((a, b.trim_end())),
+                _ => bail!(
+                    "invalid output in {}: `{}`\n\
+                    Expected a line with `{}{}KEY=VALUE` with an `=` character, \
+                    but none was found.\n\
+                    {}",
+                    whence,
+                    line,
+                    if new_syntax { "cargo::" } else { "cargo:" },
+                    if has_metadata_prefix { "metadata=" } else { "" },
+                    DOCS_LINK_SUGGESTION
+                ),
+            }
+        }
+
         for line in input.split(|b| *b == b'\n') {
             let line = match str::from_utf8(line) {
                 Ok(line) => line.trim(),
                 Err(..) => continue,
             };
-            let mut iter = line.splitn(2, ':');
-            if iter.next() != Some("cargo") {
-                // skip this line since it doesn't start with "cargo:"
-                continue;
+            let mut new_syntax = false;
+            let (key, value) = if let Some(data) = line.strip_prefix("cargo::") {
+                new_syntax = true;
+
+                // For instance, `cargo::rustc-flags=foo`.
+                if RESERVED_PREFIXES
+                    .iter()
+                    .any(|prefix| data.starts_with(prefix))
+                {
+                    parse_directive(whence.as_str(), line, data, new_syntax, false)?
+                }
+                // For instance, `cargo::metadata=foo=bar`.
+                else if let Some(data) = data.strip_prefix("metadata=") {
+                    // Reserved keys should not be used with the `cargo::metadata=` syntax.
+                    if let Some(prefix) = RESERVED_PREFIXES
+                        .iter()
+                        .find(|prefix| data.starts_with(*prefix))
+                    {
+                        bail!("invalid output in {}: `{}`\n\
+                            The reserved key cannot be used in the `cargo::metadata=` syntax. Please use `cargo::{}VAlUE` instead.\n\
+                            {}", whence, line, prefix, DOCS_LINK_SUGGESTION
+                        )
+                    }
+                    parse_directive(whence.as_str(), line, data, new_syntax, true)?
+                } else {
+                    bail!("invalid output in {}: `{}`\n\
+                        Expected a line with `cargo::metadata=KEY=VALUE` but it did not have the `metadata=` part.\n\
+                        {}", whence, line, DOCS_LINK_SUGGESTION
+                    )
+                }
```

Comment:
What is the purpose of this error case?  Are we erroring if a `RESERVED_PREFIXES` is not used?

`RESERVED_PREFIXES` is meant to only contain the original syntaxes prefixes and we should not need to update it as we add new directives (and the name should be more specific to convey that).  `match key` should be handling errors from unknown keys so we have a single source of truth

Reply:
For example, this will help with the `cargo::unknown=a` input.